### PR TITLE
refactor(components/buttonbase): convert to a functional component

### PR DESCRIPTION
**Pull request by matthewmueller**
**User-Facing Changes**

None

**Description**

Refactored the ButtonBase component from a class-based React component to a hook-based functional React component.

**References**

Part of https://github.com/foxglove/studio/issues/1426

### CODE REVIEW

File: packages/studio-base/src/components/ButtonBase.tsx [hunk 1]
```diff
@@ -11,10 +11,11 @@
 //   found at http://www.apache.org/licenses/LICENSE-2.0
 //   You may not use this file except in compliance with the License.
 
+import { makeStyles } from "@fluentui/react";
 import cx from "classnames";
-import { CSSProperties } from "react";
+import { CSSProperties, useCallback } from "react";
 
-import { LegacyButton } from "@foxglove/studio-base/components/LegacyStyledComponents";
+import { fonts, spacing } from "@foxglove/studio-base/util/sharedStyleConstants";
```
Comment (matthewmueller) [inline packages/studio-base/src/components/ButtonBase.tsx:18]: Should these be part of @fluentui as well?

Comment (defunctzombie) [reply packages/studio-base/src/components/ButtonBase.tsx:18]: I don't think fonts can be but spacing maybe could be. There is a spacing property of the fluentui theme.

**Comment by defunctzombie** [comment]
LGTM - do we still use LegacyButton? Where is ButtonBase used? Wondering if it is even necessary rather than using FluentUI buttons.

**Comment by matthewmueller** [comment]
Hey, so Legacy Button is still being used in a handful of places. It seems like quite a few of the smaller buttons.

```
packages/studio-base/src/components/LegacyStyledComponents.tsx
10: * @deprecated The LegacyButton should not be used for new features. use fluentui/react instead
12:export const LegacyButton = styled.button`

packages/studio-base/src/panels/NodePlayground/BottomBar/LogsSection.tsx
18:import { LegacyButton } from "@foxglove/studio-base/components/LegacyStyledComponents";
62:      <LegacyButton
72:      </LegacyButton>

packages/studio-base/src/components/Dropdown/index.tsx
19:import { LegacyButton } from "@foxglove/studio-base/components/LegacyStyledComponents";
132:      <LegacyButton
141:      </LegacyButton>

packages/studio-base/src/panels/Log/LogList.stories.tsx
16:import { LegacyButton } from "@foxglove/studio-base/components/LegacyStyledComponents";
100:        <LegacyButton onClick={this.togglePause}>
102:        </LegacyButton>

packages/studio-base/src/panels/Table/Table.tsx
32:  LegacyButton,
244:          <LegacyButton onClick={() => gotoPage(0)} disabled={!canPreviousPage}>
246:          </LegacyButton>{" "}
247:          <LegacyButton onClick={() => previousPage()} disabled={!canPreviousPage}>
249:          </LegacyButton>{" "}
256:          <LegacyButton onClick={() => nextPage()} disabled={!canNextPage}>
258:          </LegacyButton>{" "}
259:          <LegacyButton onClick={() => gotoPage(pageCount - 1)} disabled={!canNextPage}>
261:          </LegacyButton>{" "}

packages/studio-base/src/panels/ImageView/index.tsx
32:import { LegacyButton } from "@foxglove/studio-base/components/LegacyStyledComponents";
196:    <LegacyButton
205:    </LegacyButton>

packages/studio-base/src/panels/ImageView/ImageCanvas.tsx
24:import { LegacyButton } from "@foxglove/studio-base/components/LegacyStyledComponents";
443:          <LegacyButton className={classes.round} onClick={zoomOut} data-panel-minus-zoom>
445:          </LegacyButton>
446:          <LegacyButton className={classes.round} onClick={zoomIn} data-panel-add-zoom>
448:          </LegacyButton>
560:      <LegacyButton
566:      </LegacyButton>
```

ButtonBase is **only** used from within the Button, so at the very least, we could collapse them into a single component:

```
packages/studio-base/src/components/Button.tsx
52:        <ButtonBase style={styleAlt} {...otherProps} {...eventHandlers} className={newClassName} />
57:    <ButtonBase style={styleAlt} {...otherProps} {...eventHandlers} className={newClassName} />
```

I'm also slightly tempted to try merging Button with the Buttons in `@fluentui`, but that's a bigger project. I'd need to figure out about feature parity:

```
packages/studio-base/src/panels/Log/LogList.tsx
147:                <Button tooltip="(shortcut: scroll down)" onClick={onResetView}>

packages/studio-base/src/panels/NodePlayground/BottomBar/index.tsx
117:        <Button

packages/studio-base/src/panels/NodePlayground/index.tsx
117:      <Button style={{ marginTop: "8px" }} onClick={() => addNewNode()}>

packages/studio-base/src/panels/Publish/index.tsx
227:        <Button

packages/studio-base/src/panels/LegacyPlot/index.tsx
456:                <Button tooltip="(shortcut: double-click)" onClick={onResetZoom}>

packages/studio-base/src/panels/Internals/index.tsx
200:        <Button isPrimary small onClick={onRecordClick} data-test="internals-record-button">
219:          <Button small onClick={downloadJSON} data-test="internals-download-button">

packages/studio-base/src/panels/ThreeDimensionalViz/CameraInfo/index.tsx
93:              <Button
160:            <Button
170:            <Button
182:            <Button

packages/studio-base/src/panels/ThreeDimensionalViz/Interactions/GlobalVariableLink/UnlinkGlobalVariable.tsx
45:      <Button
62:      <Button onClick={() => setIsOpen(false)}>Cancel</Button>

packages/studio-base/src/panels/ThreeDimensionalViz/Interactions/GlobalVariableLink/LinkToGlobalVariable.tsx
108:          <Button primary={name.length > 0} disabled={name.length === 0} onClick={addLink}>
111:          <Button onClick={() => setIsOpen(false)}>Cancel</Button>

packages/studio-base/src/panels/ThreeDimensionalViz/Interactions/GlobalVariableLink/UnlinkGlobalVariables.tsx
92:            <Button

packages/studio-base/src/panels/StateTransitions/index.tsx
372:        <Button

packages/studio-base/src/components/ExtensionDetails.tsx
125:    <Button style={{ minWidth: "100px" }} onClick={onClick}>
133:    <Button style={{ minWidth: "100px" }} onClick={onClick}>

packages/studio-base/src/components/Panel.tsx
686:                <Button className={classes.tabActionsOverlayButton} onClick={groupPanels}>
692:                <Button className={classes.tabActionsOverlayButton} onClick={createTabs}>
716:                    <Button className={classes.quickActionsOverlayButton} onClick={removePanel}>
720:                    <Button className={classes.quickActionsOverlayButton} onClick={splitPanel}>

packages/studio-base/src/components/Button.tsx
52:        <ButtonBase style={styleAlt} {...otherProps} {...eventHandlers} className={newClassName} />
57:    <ButtonBase style={styleAlt} {...otherProps} {...eventHandlers} className={newClassName} />

packages/studio-base/src/components/ExtensionsSidebar/index.tsx
232:      <Button

packages/studio-base/src/components/ButtonBase.test.tsx
21:describe("<Button />", () => {
26:        <Button onClick={() => done()}>hello</Button>
36:        <Button onMouseUp={() => done()}>hello</Button>
46:        <Button onMouseLeave={() => done()}>hello</Button>
56:        <Button onFocus={() => done()}>hello</Button>
65:        <Button className="foo" onClick={done}>
78:        <Button id="button-1">hello</Button>
88:        <Button small primary warning danger>

packages/studio-base/src/components/TimeBasedChart/index.tsx
853:              <Button tooltip="(shortcut: double-click)" onClick={onResetZoom}>
```

I'd probably suggest merging this and then discussing other potential improvements. What do you think?

File: packages/studio-base/src/components/ButtonBase.tsx (hunk 1 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/ButtonBase.tsx:18]: Spacing in sharedStyleConstants is em-based, so I'd need to double-check font-size usage across the buttons and make sure there's a space for that. Open to doing that, but first I'd like to make sure we don't want to refactor the buttons anyway (see below).

**Comment by matthewmueller** [comment]
So a couple options:
1. Merge PR as is.
2. Combine `<Button>` and `<ButtonBase>`
3. Work to deprecate `<Button>` (and `<LegacyButton>`) with @fluentui's buttons.

**Comment by defunctzombie** [comment]
@matthewmueller I say merge as is. We aren't even sure if we are sticking with FluentUI medium/long term but having this as a functional component does let us make any future styling/hooks transitions easier.

File: packages/studio-base/src/components/ButtonBase.tsx [hunk 2]
```diff
@@ -35,93 +36,131 @@ export type Props = {
   innerRef?: React.Ref<HTMLButtonElement>;
 };
 
-class ButtonBaseImpl extends React.Component<Props> {
-  animationId?: ReturnType<typeof requestAnimationFrame>;
-  cancelTimeoutId?: ReturnType<typeof setTimeout>;
+const useStyles = makeStyles((theme) => ({
+  root: {
```
Comment (jtbandes) [inline packages/studio-base/src/components/ButtonBase.tsx:40]: What's the point of copying these styles from LegacyButton as opposed to just using LegacyButton directly?

Comment (matthewmueller) [reply packages/studio-base/src/components/ButtonBase.tsx]: Good catch. That was leftover from trying to remove LegacyButton altogether.

**Comment by matthewmueller** [comment]
Sorry for the delay on this one. Made a couple changes:

- Removed redundant styles
- Merged `<ButtonBase>` into `<Button>`

**Comment by 2metres** [comment]
LGTM

File: packages/studio-base/src/components/Button.tsx [hunk 3]
```diff
@@ -12,48 +12,128 @@
...
+    // replace disabled={true} with className="disabled" in order to allow
+    // tooltips on disabled buttons
+    disabled,
+    // support some bulma classes to be supplied in consumer either through
+    // bulma or custom classes these provide backwards compatibility with Studio
+    "is-small": small,
+    "is-large": large,
+    "is-primary": primary,
+    "is-warning": warning,
+    "is-danger": danger,
```
Comment (2metres) [inline packages/studio-base/src/components/Button.tsx:109]: Pretty sure these classes can be deteted as well.

**Pull request merged**